File: src/internal.ts
const replicatedStorage = game.GetService("ReplicatedStorage");
const runService = game.GetService("RunService");

export interface RequestCounter {
	Increment(player: Player): void;
	Get(player: Player): number;
}

/** @internal */
export const IS_SERVER = runService.IsServer();

/** @internal */
export const IS_CLIENT = (__LEMUR__ && !runService.IsServer()) || runService.IsClient();

/** @internal */
export const MAX_CLIENT_WAITFORCHILD_TIMEOUT = 10;

/** @internal */
export function getGlobalRemote(name: string) {
	return `$${name}`;
}

/** @internal */
export function isLuaTable(value: unknown): value is Map<unknown, unknown> {
	return typeIs(value, "table");
}

const REMOTES_FOLDER_NAME = "Remotes";
const FUNCTIONS_FOLDER_NAME = "Functions";
const EVENTS_FOLDER_NAME = "Events";

let remoteFolder: Folder;
let eventFolder: Folder;
let functionFolder: Folder;

/** @internal */
export const ServerTickFunctions = new Array<() => void>();

/** @internal */
export function findOrCreateFolder(parent: Instance, name: string): Folder {
	let folder = parent.FindFirstChild<Folder>(name);
	if (folder) {
		return folder;
	} else {
		folder = new Instance("Folder", parent);
		folder.Name = name;
		return folder;
	}
}

remoteFolder = findOrCreateFolder(replicatedStorage, REMOTES_FOLDER_NAME);
functionFolder = findOrCreateFolder(remoteFolder, FUNCTIONS_FOLDER_NAME);
eventFolder = findOrCreateFolder(remoteFolder, EVENTS_FOLDER_NAME);

/**
 * Errors with variables formatted in a message
 * @param message The message
 * @param vars variables to pass to the error message
 */
export function errorft(message: string, vars: { [name: string]: unknown }) {
	message = message.gsub("{([%w_][%w%d_]*)}", (token: string) => {
		return vars[token] || token;
	});

	error(message, 2);
}

/** @internal */
export function eventExists(name: string) {
	return eventFolder.FindFirstChild(name) !== undefined;
}

/** @internal */
export function functionExists(name: string) {
	return functionFolder.FindFirstChild(name) !== undefined;
}

/** @internal */
export function waitForEvent(name: string, timeOut: number): RemoteEvent | undefined {
	return eventFolder.WaitForChild(name, timeOut);
}

/** @internal */
export function waitForFunction(name: string, timeOut: number): RemoteFunction | undefined {
	return functionFolder.WaitForChild(name, timeOut);
}

/** @internal */
export function getRemoteFolder<K extends keyof RemoteTypes>(remoteType: K): Folder {
	let targetFolder: Folder;
	if (remoteType === "RemoteEvent") {
		targetFolder = eventFolder;
	} else if (remoteType === "RemoteFunction") {
		targetFolder = functionFolder;
	} else {
		return error("Invalid type: " + remoteType);
	}

	return targetFolder;
}

/** @internal */
export function findRemote<K extends keyof RemoteTypes>(remoteType: K, name: string): RemoteTypes[K] | undefined {
	const targetFolder = getRemoteFolder(remoteType);
	const existing = targetFolder.FindFirstChild(name) as RemoteFunction | RemoteEvent;

	return existing as RemoteTypes[K] | undefined;
}

/** @internal */
export function getRemoteOrThrow<K extends keyof RemoteTypes>(remoteType: K, name: string): RemoteTypes[K] {
	const existing = findRemote(remoteType, name);
	if (existing) {
		return existing;
	} else {
		throw `Could not find Remote of type ${remoteType} called "${name}"`;
	}
}

/** @internal */
export function findOrCreateRemote<K extends keyof RemoteTypes>(remoteType: K, name: string): RemoteTypes[K] {
	const existing = findRemote(remoteType, name);
	if (existing) {
		return existing;
	} else {
		if (!IS_SERVER) {
			throw "Creation of Events or Functions must be done on server!";
		}

		let remote: RemoteEvent | RemoteFunction;

		if (remoteType === "RemoteEvent" || remoteType === "RemoteFunction") {
			remote = new Instance(remoteType);
		} else {
			throw `Invalid Remote Type: ${remoteType}`;
		} // stfu

		remote.Name = name;
		remote.Parent = getRemoteFolder(remoteType);
		return remote as RemoteTypes[K];
	}
}

if (IS_SERVER) {
	game.GetService("RunService").Stepped.Connect((time, step) => {
		for (const f of ServerTickFunctions) {
			f();
		}
	});
}

File: internal.lua

local replicatedStorage = game:GetService("ReplicatedStorage")
local runService = game:GetService("RunService")

local IS_SERVER = runService:IsServer()

local IS_CLIENT = (__LEMUR__ and not runService:IsServer()) or runService:IsClient()

local MAX_CLIENT_WAITFORCHILD_TIMEOUT = 10

local function getGlobalRemote(name)
	return "$" .. name
end

local function isLuaTable(value)
	local _0 = value
	return type(_0) == "table"
end
local REMOTES_FOLDER_NAME = "Remotes"
local FUNCTIONS_FOLDER_NAME = "Functions"
local EVENTS_FOLDER_NAME = "Events"
local ASYNC_FUNCTIONS_FOLDER_NAME = "AsyncFunctions"

local ServerTickFunctions = {}

local function findOrCreateFolder(parent, name)
	local folder = parent:FindFirstChild(name)
	if folder then
		return folder
	else
		folder = Instance.new("Folder", parent)
		folder.Name = name
		return folder
	end
end
local remoteFolder = findOrCreateFolder(replicatedStorage, REMOTES_FOLDER_NAME)
local functionFolder = findOrCreateFolder(remoteFolder, FUNCTIONS_FOLDER_NAME)
local eventFolder = findOrCreateFolder(remoteFolder, EVENTS_FOLDER_NAME)
local asyncFunctionFolder = findOrCreateFolder(remoteFolder, ASYNC_FUNCTIONS_FOLDER_NAME)
--[[
	*
	* Errors with variables formatted in a message
	* @param message The message
	* @param vars variables to pass to the error message
]]
local function errorft(message, vars)
	
	
	local _0 = message
	local _1 = function(token)
		local _2 = vars[token]
		if not (_2 ~= 0 and _2 == _2 and _2 ~= "" and _2) then
			_2 = token
		end
		return _2
	end
	message = string.gsub(_0, "{([%w_][%w%d_]*)}", _1)
	error(message, 2)
end

local function eventExists(name)
	return eventFolder:FindFirstChild(name) ~= nil
end

local function functionExists(name)
	return functionFolder:FindFirstChild(name) ~= nil
end

local function waitForEvent(name, timeOut)
	return eventFolder:WaitForChild(name, timeOut)
end

local function waitForFunction(name, timeOut)
	return functionFolder:WaitForChild(name, timeOut)
end

local function getRemoteFolder(remoteType)
	local targetFolder
	if remoteType == "RemoteEvent" then
		targetFolder = eventFolder
	elseif remoteType == "RemoteFunction" then
		targetFolder = functionFolder
	elseif remoteType == "AsyncRemoteFunction" then
		targetFolder = asyncFunctionFolder
	else
		return error("Invalid type: " .. remoteType)
	end
	return targetFolder
end

local function findRemote(remoteType, name)
	local targetFolder = getRemoteFolder(remoteType)
	local existing = targetFolder:FindFirstChild(name)
	return existing
end

local function getRemoteOrThrow(remoteType, name)
	local existing = findRemote(remoteType, name)
	if existing then
		return existing
	else
		error("Could not find Remote of type " .. remoteType .. ' called "' .. name .. '"')
	end
end

local function findOrCreateRemote(remoteType, name)
	local existing = findRemote(remoteType, name)
	if existing then
		return existing
	else
		if not IS_SERVER then
			error("Creation of Events or Functions must be done on server!")
		end
		local remote
		if remoteType == "RemoteEvent" then
			remote = Instance.new("RemoteEvent")
		elseif remoteType == "AsyncRemoteFunction" then
			remote = Instance.new("RemoteEvent")
		elseif remoteType == "RemoteFunction" then
			remote = Instance.new("RemoteFunction")
		else
			error("Invalid Remote Type: " .. remoteType)
		end
		remote.Name = name
		remote.Parent = getRemoteFolder(remoteType)
		return remote
	end
end
local function checkArguments(types, args)
	if args == nil then
		warn("[net-types] Argument length is zero")
		return false
	end
	do
		local _0 = 0
		while _0 < #types do
			local i = _0
			local typeCheck = types[i + 1]
			local value = args[i + 1]
			if not typeCheck(value) then
				warn("[net-types] Argument at index " .. tostring(i) .. " was invalid type.")
				return false
			end
			_0 = i
			_0 += 1
		end
	end
	return true
end
if IS_SERVER then
	game:GetService("RunService").Stepped:Connect(function(time, step)
		for _, f in ipairs(ServerTickFunctions) do
			f()
		end
	end)
end
return {
	getGlobalRemote = getGlobalRemote,
	isLuaTable = isLuaTable,
	findOrCreateFolder = findOrCreateFolder,
	errorft = errorft,
	eventExists = eventExists,
	functionExists = functionExists,
	waitForEvent = waitForEvent,
	waitForFunction = waitForFunction,
	getRemoteFolder = getRemoteFolder,
	findRemote = findRemote,
	getRemoteOrThrow = getRemoteOrThrow,
	findOrCreateRemote = findOrCreateRemote,
	checkArguments = checkArguments,
	IS_SERVER = IS_SERVER,
	IS_CLIENT = IS_CLIENT,
	MAX_CLIENT_WAITFORCHILD_TIMEOUT = MAX_CLIENT_WAITFORCHILD_TIMEOUT,
	ServerTickFunctions = ServerTickFunctions,
}
